### PR TITLE
#128 Added functions to retrieve the current trainees and IPs

### LINKED ISSUE #128: Server: Implement Traineesection (not-admin)
The server provides the following data:
- For every trainee generation all internal projects
  - ID
  - Name 
  - Kick-Off date
  - ZP date
  -  AP date
  - "Angebot" submitted (boolean)
  - "Dokumentationsleitfaden" (DL) submitted (boolean)
- For every trainee generation all trainees:
  - ID
  - Name (firstname and lastname)

Data can be accessed by every logged in user.

_See page 51 development plan_

### PR DESCRIPTION
**Pull request by jbt-it**
The folder trainees for the trainees module was added as well as the respective files for the controller, routes and types. In addtion the lines of code were added to the app.ts file so the routes can be used.

For this Issue two functions and routes were added: the `retrieveCurrentTrainees` and `retrieveCurrentIPs` function and their respective routes "/" and "/currentIPs".

The function `retrieveCurrentTrainees` is supposed to "get the data to display the trainee generation". By checking the URL GET https://localhost:3030/trainees with Postman a JSON-object with the current trainees should be received if logged in.

The function `retrieveCurrentIPs` is supposed to "get the data to display the IPs". By checking the URL GET https://localhost:3030/trainees/currentIPs with Postman a JSON-object with the current trainees should be received if logged in.

Closes #128

### CODE REVIEW

File: server/src/trainees/traineesController.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ */
+export const retrieveTraineeGenerations = (req: Request, res: Response): void => {
+    //numberOfGenerations = database.query(SELECT MAX(generationID) FROM generation)
+    
+    let numberOfGenerations: number;
+    database.query('SELECT MAX(generationID) FROM generation',[]).then((result: number)=>{
+        numberOfGenerations = result;
+    });
+ 
+    for (var i = 1; i< numberOfGenerations + 1;i++){
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:69]: use `let` instead of `var`.

File: server/src/trainees/traineesController.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+
+
+  /**
+ * Retrieves all trainee generations
+ * SELECT MAX(generationID) FROM generation; 
+ * For all i to Max(generationID) create type?
+ */
+export const retrieveTraineeGenerations = (req: Request, res: Response): void => {
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:61]: There is no route for this controller function.
How should this function be tested?

**Review by LennartRukower — CHANGES_REQUESTED** [review]
A few minor issues and one question about the third implemented function.

Please check if you have prettier and TSLint activated because there are a few formatting issues in your files.

File: client/src/ts/global/App.tsx [hunk 3]
```diff
@@ -63,6 +63,7 @@ const App: React.FunctionComponent = () => {
         setCheckAuthLoading(false);
       })
       .catch((err) => {
+        console.log(err);
```
Comment (LennartRukower) [inline client/src/ts/global/App.tsx:66]: Is this console log necessary? Why was this added?

File: server/src/trainees/traineesController.ts [hunk 4]
```diff
@@ -55,3 +55,66 @@ export const retrieveCurrentIPs = (req: Request, res: Response): void => {
...
+  database
+    .query(
+      `SELECT *
+      FROM internesprojekt 
+      ORDER BY generation`,
+      []
+    )
+    .then((result: traineesTypes.GetAllIPsQueryResult[]) => {
+      if (result.length === 0) {
+        res.status(404).send("IPs not found");
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:72]: 404 should only be used, if a specific resource is not found.
E.g. If you want to retrieve an IP with id 66 and this IP does not exist, a 404 would be thrown.
If we retrieve all IPs and there are no IPs we do not need to send an error message. The user gets an empty array, which is sufficient and better for handling on the frontend.

File: server/src/trainees/traineesController.ts [hunk 5]
```diff
@@ -55,3 +55,66 @@ export const retrieveCurrentIPs = (req: Request, res: Response): void => {
...
+    .catch((err) => {
+      res.status(500).send("Query Error");
+    });
+};
+
+/*
+  Retrieve a list of all Trainees with their respective IPs
+  Group all Trainees into their corresponding generations
+*/
+export const retrieveAllTrainees = (req: Request, res: Response): void => {
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:87]: Maybe call this function `retrieveAllTraineesWithIPs`

File: server/src/trainees/traineesTypes.ts [hunk 6]
```diff
@@ -2,44 +2,56 @@
...
+  DLbeiEV: boolean;
+};
 
-    //...
-  };
\ No newline at end of file
+/**
+ * Type of query results of trainee generations retrieval.
+ */
+export type retrieveAllTraineesQueryResult = {
```
Comment (LennartRukower) [inline server/src/trainees/traineesTypes.ts:51]: Call this type `GetAllTraineesWithIPsQueryResult`

File: server/src/trainees/traineesController.ts [hunk 7]
```diff
@@ -55,3 +55,66 @@ export const retrieveCurrentIPs = (req: Request, res: Response): void => {
...
+    .query(
+      `SELECT mitglied.mitgliedID, mitglied.vorname, mitglied.nachname, mitglied.generation, internesprojekt.internesprojektID, internesprojekt.projektname, internesprojekt.kuerzel, internesprojekt.kickoff, internesprojekt.AngebotBeiEV, internesprojekt.ZPbeiEV, internesprojekt.ZPgehalten, internesprojekt.APbeiEV, internesprojekt.APgehalten, internesprojekt.DLbeiEV
+      FROM mitglied
+      LEFT JOIN internesprojekt ON mitglied.generation = internesprojekt.generation 
+      ORDER BY generation`,
+      []
+    )
+    .then((result: traineesTypes.retrieveAllTraineesQueryResult[]) => {
+      if (result.length === 0) {
+        res.status(404).send("Trainees not found");
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:98]: 404 should only be used, if a specific resource is not found.
E.g. If you want to retrieve an IP with id 66 and this IP does not exist, a 404 would be thrown.
If we retrieve all IPs and there are no IPs we do not need to send an error message. The user gets an empty array, which is sufficient and better for handling on the frontend.

File: server/src/trainees/traineesController.ts [hunk 8]
```diff
@@ -55,3 +55,66 @@ export const retrieveCurrentIPs = (req: Request, res: Response): void => {
...
+export const retrieveAllTrainees = (req: Request, res: Response): void => {
+  database
+    .query(
+      `SELECT mitglied.mitgliedID, mitglied.vorname, mitglied.nachname, mitglied.generation, internesprojekt.internesprojektID, internesprojekt.projektname, internesprojekt.kuerzel, internesprojekt.kickoff, internesprojekt.AngebotBeiEV, internesprojekt.ZPbeiEV, internesprojekt.ZPgehalten, internesprojekt.APbeiEV, internesprojekt.APgehalten, internesprojekt.DLbeiEV
+      FROM mitglied
+      LEFT JOIN internesprojekt ON mitglied.generation = internesprojekt.generation 
+      ORDER BY generation`,
+      []
+    )
+    .then((result: traineesTypes.retrieveAllTraineesQueryResult[]) => {
```
Comment (LennartRukower) [inline server/src/trainees/traineesController.ts:96]: Is `traineesTypes.retrieveAllTraineesQueryResult[]` the correct type for this query result?
`export type retrieveAllTraineesQueryResult = {
  generation: number;
  mitgliedID: number;
  vorname: string;
  nachname: string;
  internesprojekt: string;
};` -> internesprojekt is a string?

File: server/src/trainees/traineesRoutes.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ 
+
+/**
+ * =======>>> ALL routes after this point are accessible for logged in users only <<<=======
+ */
+ router.use(authController.protectRoutes);
+ 
+ //  =======>>> Get routes <<<=======
+ router.get("/", traineesController.retrieveCurrentTrainees);
+ router.get("/currentIPs", traineesController.retrieveCurrentIPs);
```
[the remark this answers is not in the document]
Comment (JanMueller31) [reply server/src/trainees/traineesRoutes.ts]: Done

File: server/src/trainees/traineesController.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+ * SELECT MAX(generationID) FROM generation; 
+ * For all i to Max(generationID) create type?
+ */
+export const retrieveTraineeGenerations = (req: Request, res: Response): void => {
+    //numberOfGenerations = database.query(SELECT MAX(generationID) FROM generation)
+    
+    let numberOfGenerations: number;
+    database.query('SELECT MAX(generationID) FROM generation',[]).then((result: number)=>{
+        numberOfGenerations = result;
+    });
```
[the remark this answers is not in the document]
Comment (JanMueller31) [reply server/src/trainees/traineesController.ts]: Changed

File: client/src/ts/global/App.tsx (hunk 3 above)
Comment (LennartRukower) [reply client/src/ts/global/App.tsx:66]:
```suggestion
```

**Review by LennartRukower — CHANGES_REQUESTED** [review]

**Review by LennartRukower — APPROVED** [review]
LGTM, nice work

**Pull request merged**